src/main/java/olya/app/remindme/service/impl/AuthServiceImpl.java

```diff
@@ -1,38 +1,75 @@
 import olya.app.remindme.security.JwtIssuer;
 import olya.app.remindme.security.UserPrincipal;
 import olya.app.remindme.service.AuthService;
+import olya.app.remindme.service.UserService;
+import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.security.authentication.AuthenticationManager;
 import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
 import org.springframework.security.core.Authentication;
-import org.springframework.security.core.GrantedAuthority;
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.stereotype.Service;
 
 @Service
+@RequiredArgsConstructor
 public class AuthServiceImpl implements AuthService {
     private final JwtIssuer jwtIssuer;
     private final AuthenticationManager authenticationManager;
-
-    public AuthServiceImpl(JwtIssuer jwtIssuer, AuthenticationManager authenticationManager) {
-        this.jwtIssuer = jwtIssuer;
-        this.authenticationManager = authenticationManager;
-    }
+    private final UserService userService;
+    private final TokenRepository tokenRepository;
 
     public LoginResponse attemptLogin(String email, String password) {
         Authentication authentication = authenticationManager.authenticate(
                 new UsernamePasswordAuthenticationToken(email, password)
         );
         SecurityContextHolder.getContext().setAuthentication(authentication);
         UserPrincipal principal = (UserPrincipal) authentication.getPrincipal();
+        Instant expiresAt = Instant.now().plus(Duration.of(1, ChronoUnit.DAYS));
 //        List<String> roles = principal.getAuthorities()
 //                .stream()
 //                .map(GrantedAuthority::getAuthority)
 //                .toList();
-        String token = jwtIssuer.issue(principal.getUserId(), principal.getEmail());
+        String token = jwtIssuer.issue(principal.getUserId(), expiresAt, principal.getEmail());
+        createTokenEntity(token, expiresAt, principal.getEmail());
         return new LoginResponse(token);
     }
+
+    public void logout(String email) {
+        revokeUserTokens(userService.getByEmail(email));
+        SecurityContextHolder.clearContext();
+    }
+
+    private void createTokenEntity(String token, Instant expiresAt, String email) {
+        Token tokenEntity = new Token();
+        tokenEntity.setToken(token);
+        tokenEntity.setRevoked(false);
+        tokenEntity.setValidUntil(Timestamp.from(expiresAt));
+        tokenEntity.setUser(userService.getByEmail(email));
+        tokenRepository.save(tokenEntity);
+    }
+
+    private void revokeUserTokens(User user) {
+        List<Token> userTokens = tokenRepository.findByUserIdAndRevokedFalse(user.getId());
+        if (userTokens.isEmpty())
+            return;
+        userTokens.forEach(t -> t.setRevoked(true));
+        tokenRepository.saveAll(userTokens);
+    }
+
+    @Scheduled(cron = "0 34 19 * * SUN")
```

Comment:
Why does this ren only on Sunday?

Reply:
I was testing it on Sunday. But there is nothing particular with a week day. Can be any or everyday (*) if we want.